src/built-in-macros.ls

```diff
@@ -188,117 +187,122 @@ contents =
     elements : elements.map @compile
 
   \object : do
-    check-list = (list, i) ->
-      | list? and list.type is \list => list.values
-      | otherwise => throw Error "Expected property #i to be a list"
-
-    infer-name = (prefix, name, computed) ->
-      if computed
-        prefix
-      else if typeof name.type is \Literal
-        "#prefix #{name.value}"
-      else
-        "#prefix #{name.name}"
-
-    compile-get-set = (i, type, [name, params, ...body]) ->
+    # This macro needs to detect patterns in its arguments, e.g.
+    #
+    #     (object (get 'a () (return 1)))
+    #
+    # where (get <something> <parameters> <body...>) is a pattern.  It is split
+    # into methods that handle different kinds of patterns.
+
+    # Specific Error type, to be thrown just in this macro when the user has
+    # provided an invalid argument pattern.
+    #
+    # This makes error handling neater: the top-level macro function catches
+    # this type of Errors and prepends information about which parameter was
+    # being processed when it occurred.
+    class ObjectParamError extends Error
```

Comment:
I would also like to mention that this really needs to be done everywhere, not just within the `(object)` macro. But I'll save that for a later exercise.

Also, please include the stack. It makes for easier debugging. (Code generation serves as a good example here.)

```ls
# This is sufficient.
if Error.captureStackTrace
  Error.captureStackTrace this, ObjectParamError
else
  @stack = new Error!stack
```